rfcs/017 - Cluster Data Storage RFC.md

```diff
@@ -0,0 +1,221 @@
+
+![A partial partition diagram of the paragraph above](./partial_partition.jpg)
+
+This all can be contained in a relatively simple `/ping/<public_ip_address>` control port endpoint. Normally the `/ping/<public_ip_address>` would be directed at `<public_ip_address>`, which should always return `ok` (as long as the cluster secret token is also provided as a header, otherwise `fail`, but that's at an earlier part of the control port implementation). But if a request for `/ping/<public_ip_address>` comes in and `<public_ip_address>` is not the node's own public IP address, then it immediately forwards the request on to the public IP address in the request URL and returns its response or `fail` if it times out (should be a short timeout as well, such as 50ms, but exact threshold depends on real world data here).
+
+The datastore node API should work similarly but at a higher level. The node determines based on LogRendezvousHash which node should own the namespace+key pair in question, then it fires that request to the *appropriate* node, which is the node in question under normal circumstances but if the internal bookkeeping of node routing has an intermediary node listed, it goes to that node instead. When handling a request for the data, the node receiving the request hashes the namespace+key pair and determines if it is the correct node for the job or if it should forward that request to another node, which will make this process seamless.
+
+This exact functionality can be automatically used by the `Anycloud` app by just requesting something like `/datastore/has/<namespace>/<key>` to the local instance and being sure that it will forward along the request to the correct node, or the local instance happens to actually own that data and handles it itself.
+
+Providing the cluster secret string into the Docker container as an environment variable along with the 4142 port is the other piece that makes it possible, and these requests can be put into a simple library for the language the Anycloud app is written in to cover up these implementation details and make it fit in better with the language in question.
+
+But those endpoints make clear something else that's necessary: the `HandlerMemory` structs that represent the value of a particular key need to be able to be serialized and deserialized, and in a way that they can be done so not just in Rust, but any language we want to fully support in Anycloud. (That's not entirely correct, we could create a set of secondary endpoints just for Anycloud that only accepts strings, or rather `Array<int8>` arrays where strings are a special case of that, but that may limit/impede future efforts at RPC between the Alan and Anycloud for offloading CPU-intensive work to Alan, and we still need serialization/deserialization of HandlerMemory for Alan itself.)
```

Comment:
I'm a bit confused. Isn't `@std/datastore` backed by a [Dashmap](https://crates.io/crates/dashmap)? Would that change?

Reply:
No, that wouldn't change. What would change is that we need to put a thin layer on top of it to:

1. Go ask another node via its control port for the data if it's not the primary node for that data.
2. Any time it writes data to the local DashMap, also push that data to the 2 other replicas in the cluster (assuming there are at least 3 nodes in the cluster).
3. Every time the node assignments have changed, ask every other node if there's data it should now be in charge of and drop any data it is no longer even in charge of replicating from its own DashMap.

However, because we discovered that `DashMap` has a huge perf hit when dropping data, it may make sense to re-eval its use since dropping keys is expected in the normal lifetime of the datastore. But I also considered that an implementation detail that doesn't belong in the RFC. The responsibilities of the thin layer *should* be in the RFC, though.

Reply:
It is an implementation detail, but I was missing the "thin layer on top of it" part. Thanks for clarifying!